linkerd/app/core/src/proxy/server.rs

```diff
@@ -92,36 +127,28 @@ where
         self.0.poll_ready().map_err(Into::into)
     }
 
-    fn call(&mut self, s: Source) -> Self::Future {
-        let target = match s.orig_dst {
-            Some(addr) => T::from(addr),
-            None => return future::Either::A(future::err(NoOriginalDst.into())),
-        };
-
-        future::Either::B(self.0.call(target).map_err(Into::into))
-    }
-}
+    fn call(&mut self, meta: tls::accept::Meta) -> Self::Future {
+        if meta.addrs.target_addr_is_local() {
+            return future::Either::A(future::err(NoForwardTarget.into()));
+        }
 
-impl<T, C: Clone> Clone for ForwardConnect<T, C> {
-    fn clone(&self) -> Self {
-        ForwardConnect(self.0.clone(), PhantomData)
+        future::Either::B(self.0.call(meta.addrs.target_addr()).map_err(Into::into))
     }
 }
 
-impl error::Error for NoOriginalDst {}
+impl error::Error for NoForwardTarget {}
 
-impl fmt::Display for NoOriginalDst {
+impl fmt::Display for NoForwardTarget {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Missing SO_ORIGINAL_DST address")
+        write!(f, "Could not forward to a target address")
```

Comment:
Looks like this is only constructed if the target address is local, it might be good if the message indicated that for logs etc.